ateam_vision_filter/src/world.cpp

```diff
@@ -0,0 +1,206 @@
+      camera_pair.second.get_ball_estimate_with_score();
+
+    if (possible_ball_with_score.has_value()) {
+      balls_with_scores.emplace_back(possible_ball_with_score.value());
+    }
+  }
+
+  // If we have no estimates, there is no ball
+  if (balls_with_scores.empty()) {
+    return std::nullopt;
+  }
+
+  Ball merged_ball;
+  double total_score = 0.0;
+  for (const auto & ball_with_score : balls_with_scores) {
+    const auto & ball = ball_with_score.first;
+    const auto & score = ball_with_score.second;
+
+    merged_ball.position += score * ball.position;
+    merged_ball.velocity += score * ball.velocity;
+    merged_ball.acceleration += score * ball.acceleration;
+    total_score += score;
+  }
+
+  merged_ball.position /= total_score;
+  merged_ball.velocity /= total_score;
+  merged_ball.acceleration /= total_score;
+
+  return merged_ball;
+}
+
+std::array<std::optional<Robot>, 16> World::get_yellow_robots_estimate()
+{
+  using RobotWithScore = std::pair<Robot, double>;
+  std::array<std::vector<RobotWithScore>, 16> yellow_robots_with_scores;
+
+  // Grab estimates from camera
+  for (auto & camera_pair : cameras) {
+    std::array<std::optional<RobotWithScore>, 16> possible_yellow_robots_with_score =
+      camera_pair.second.get_yellow_robot_estimates_with_score();
+
+    for (size_t yellow_id = 0; yellow_id < 16; yellow_id++) {
+      if (possible_yellow_robots_with_score.at(yellow_id).has_value()) {
+        yellow_robots_with_scores.at(yellow_id).emplace_back(
+          possible_yellow_robots_with_score.at(yellow_id).value());
+      }
+    }
+  }
+
+  // Try to merge them together
+  // return nullopt if there are no measurements
+  std::array<std::optional<Robot>, 16> yellow_robots_estimates;
+  for (size_t yellow_id = 0; yellow_id < 16; yellow_id++) {
+    if (yellow_robots_with_scores.empty()) {
+      yellow_robots_estimates.at(yellow_id) = std::nullopt;
+    } else {
+      // Merge robots based on weighted average of their scores
+      yellow_robots_estimates.at(yellow_id) = Robot();
+
+      auto & output_robot = yellow_robots_estimates.at(yellow_id).value();
+      Eigen::Vector2d output_angle{0, 0};
+      double total_score = 0.0;
+
+      for (const auto & robot_with_score : yellow_robots_with_scores.at(yellow_id)) {
+        const Robot & robot = std::get<0>(robot_with_score);
+        const double & score = std::get<1>(robot_with_score);
+        output_robot.position += score * robot.position;
+        output_angle += score * Eigen::Vector2d{std::cos(output_robot.theta), std::sin(
+            output_robot.theta)};
+        output_robot.velocity += score * robot.velocity;
+        output_robot.omega += score * robot.omega;
+        output_robot.acceleration += score * robot.acceleration;
+        output_robot.alpha += score * robot.alpha;
+
+        total_score += score;
+      }
+
+      output_robot.position /= total_score;
+      output_robot.theta = 0.0;
+      if (output_angle.norm() > 0) {

```

Comment:
There's an [isZero()](https://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#af36014ec300f53a65083057ed4e89822) function.
```suggestion
      if (!output_angle.isZero()) {
```